velox/exec/Task.cpp

```diff
@@ -1442,6 +1464,220 @@ Task::getLocalExchangeQueues(
+  constexpr int64_t kMinRecoverableBytes = 10 << 20;
+
+  // When running with memory arbitration, the initial reservation comes from
+  // the first action of the first Driver of the new Task. Therefore this is
+  // rounded up to at least 24MB per Task.
+  constexpr int64_t kInitialSize = 24 << 20;
+  std::lock_guard<std::mutex> l(mutex_);
+  Task* consumerTask = dynamic_cast<Task*>(requester.get());
+  VELOX_CHECK(consumerTask, "Only a Task can request memory via recover()");
+  auto topTracker =
+      memory::getProcessDefaultMemoryManager().getMemoryUsageTracker();
+  auto& tracker = consumerTask->tracker();
+  // The limits of the consumers are stable inside this section but
+  // the allocation sizes are volatile until the consumer in question
+  // is stopped.
+
+  if (consumerTask->state() != kRunning) {
+    return false;
+  }
+  int64_t bytesForRequester =
+      tracker.maxTotalBytes() < kMinRecoverableBytes ? kInitialSize : bytes;
+  auto available =
+      topTracker->maxTotalBytes() - topTracker->getCurrentUserBytes();
+
+  // Allocation can be made from unallocated memory.
+  if (bytesForRequester <= available) {
+    if (maybeUpdate(bytesForRequester, *topTracker)) {
+      setLimit(tracker.maxTotalBytes() + bytesForRequester, tracker);
+      VLOG(1) << fmt::format(
+          "{} gets {} from unused", consumerTask->taskId(), bytesForRequester);
+      return true;
+    }
+  }
+
+  std::vector<ConsumerScore> candidates;
+  available = 0;
+  for (auto [raw, weak] : consumers_) {
+    auto ptr = weak.lock();
+    if (ptr) {
+      auto otherTask = dynamic_cast<Task*>(ptr.get());
+      if (otherTask) {
+        auto size = otherTask->tracker().getCurrentTotalBytes();
+        if (size > kInitialSize) {
+          auto recoverable = otherTask->recoverableMemory();
+          if (recoverable >= kMinRecoverableBytes) {
+            candidates.push_back({ptr, otherTask, recoverable});
+            available += candidates.back().available;
+          }
+        }
+      }
+    }
+  }
+  if (available < bytesForRequester) {
+    return false;
+  }
+  std::sort(
+      candidates.begin(),
+      candidates.end(),
+      [&](const ConsumerScore& left, const ConsumerScore& right) {
+        // Most available first.
+        return left.available > right.available;
+      });
+  int64_t recovered = 0;
+  std::vector<std::shared_ptr<Task>> pausedTasks;
+  for (auto& candidate : candidates) {
+    if (auto& task = candidate.task) {
+      auto future = task->requestPause(true);
+      auto& exec = folly::QueuedImmediateExecutor::instance();
+      std::move(future).via(&exec).wait();
+      if (task->state() != TaskState::kRunning) {
+        continue;
+      }
+      VELOX_CHECK_EQ(0, task->numThreads());
+      pausedTasks.push_back(std::static_pointer_cast<Task>(candidate.consumer));
+
+      auto& taskTracker = task->tracker();
+      auto previousBytes = taskTracker.maxTotalBytes();
+      task->recover(bytesForRequester - recovered);
```

Comment:
Ok. The algorithm here is to reclaim as much as possible from the first candidate which has most revocable memory? Thanks!